src/main/java/com/mongodb/kafka/connect/source/heartbeat/HeartbeatConsumer.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.kafka.connect.source.heartbeat;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.apache.kafka.clients.consumer.CommitFailedException;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.errors.WakeupException;
+import org.apache.kafka.common.utils.Bytes;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class HeartbeatConsumer implements Runnable {
+  private static final Logger LOGGER = LoggerFactory.getLogger(HeartbeatConsumer.class);
+  private static final String CONSUMER_GROUP_ID = "MONGODB_SOURCE_HEARTBEAT";
+  private static final String CONSUMER_DESERIALIZER =
+      "org.apache.kafka.common.serialization.BytesDeserializer";
+
+  private final AtomicBoolean running;
+  private final KafkaConsumer<Bytes, Bytes> consumer;
+  private final String heartbeatTopicName;
+  private final long heartbeatIntervalMS;
+
+  HeartbeatConsumer(
+      final List<String> bootStrapServers,
+      final long heartbeatIntervalMS,
+      final String heartbeatTopicName) {
+    this.consumer = tryCreateConsumer(String.join(",", bootStrapServers)).orElse(null);
+    this.heartbeatIntervalMS = heartbeatIntervalMS;
+    this.heartbeatTopicName = heartbeatTopicName;
+    this.running = new AtomicBoolean(false);
+    if (consumer != null) {
+      new Thread(this).start();
```

Comment:
Roughly speaking, passing `this` to a different thread `t` (regardless of the way it is passed) abolishes the [`final` fields semantics](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5.1)<sup>1</sup> for reads from `final` fields of `this` in `t` and any other thread that `t` shares a reference to `this` with.

Here is what JSL [says](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5) about this in plain text:

> "An object is considered to be _completely initialized_ when its constructor finishes. A thread that can only see a reference to an object after that object has been completely initialized is guaranteed to see the correctly initialized values for that object's `final` fields.
> The usage model for `final` fields is a simple one: Set the `final` fields for an object in that object's constructor; and do not write a reference to the object being constructed in a place where another thread can see it before<sup>2</sup> the object's constructor is finished. If this is followed, then when the object is seen by another thread, that thread will always see the correctly constructed version of that object's `final` fields. It will also see versions of any object or array referenced by those `final` fields that are at least as up-to-date as the `final` fields are."

Here is the same [expressed formally](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5) ( ͡° ͜ʖ ͡°):

> Given a write `w`, a freeze `f`, an action `a` (that is not a read of a `final` field), a read `r1` of the `final` field frozen by `f`, and a read `r2` such that `hb(w, f)`, `hb(f, a)`, `mc(a, r1)`, and `dereferences(r1, r2)`, then when determining which values can be seen by `r2`, we consider `hb(w, r2)`. (This happens-before ordering does not transitively close with other happens-before orderings.)

So we need to have a completely initialized `HeartbeatConsumer` before we can pass it to either existing or a new thread ([daemon](https://docs.oracle.com/javase/8/docs/api/java/lang/Thread.html#setDaemon-boolean-) created via [`Executors.defaultThreadFactory`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Executors.html#defaultThreadFactory--) to allow users at least some control over the threads a connector starts?). Also, `HeartbeatConsumer` does not need to be `Runnable` (`() -> run()` may help), in which case it could have a private `run` method, from which IDE can easily find usages (searching for usages of `Runnable.run` is terrifying and eventually makes my IntelliJ IDEA unusable).

I also would like to point out that this rises an additional<sup>3</sup> concern about the code that does IO from constructors. For asynchronous IO this may result in not completely initialized `this` leaking to a different thread.

----
<sup>1</sup> This semantics is what differentiates [`final` fields](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5) from [`final` variables](https://docs.oracle.com/javase/specs/jls/se8/html/jls-4.html#jls-4.12.4) and allows programmers to [_"implement thread-safe immutable objects without synchronization"_](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.5). I may recommend the following [links to useful (trustworthy) resources about the JMM](https://www.kovalenko.link/blog/java-final-field-semantics#links) I gathered, if reading [JMM](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.4) is considered boring.
<sup>2</sup> This "before" is informal here because it cannot be expressed as a simple yet formally correct rule.
<sup>3</sup> Additional to a concern of an object doing observable actions without having anyone requested any activity.